lnwallet/rpcwallet/remote_signer_client.go

```diff
@@ -0,0 +1,939 @@
+		// returned yet.
+		select {
+		case <-ctx.Done():
+			return
+
+		case connDoneChan <- conn:
+		}
+	}()
+
+	select {
+	case conn := <-connDoneChan:
+		return conn, nil
+
+	case err := <-errChan:
+		return nil, err
+
+	case <-s.quit:
+		return nil, ErrShuttingDown
+
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+// A compile time assertion to ensure StreamFeeder meets the
+// SignCoordinatorStreamFeeder interface.
+var _ SignCoordinatorStreamFeeder = (*StreamFeeder)(nil)
+
+// RemoteSignerClient is a client which will process and respond to sign
+// requests from the watch-only node, which are sent over a stream between the
+// node and a watch-only node.
+//
+// Note: The client will only start fully if the passed configuration enables an
+// outbound remote signer.
+type RemoteSignerClient struct {
+	// walletServer is the WalletKitServer that the remote signer client
+	// will use to process walletrpc requests.
+	walletServer walletrpc.WalletKitServer
+
+	// signerServer is the SignerServer that the remote signer client will
+	// use to process signrpc requests.
+	signerServer signrpc.SignerServer
+
+	// streamFeeder is the stream feeder that will set up a stream to the
+	// watch-only node when requested to do so by the remote signer client.
+	streamFeeder SignCoordinatorStreamFeeder
+
+	// stream is the stream between the node and the watch-only node.
+	stream walletrpc.WalletKit_SignCoordinatorStreamsClient
+
+	// shouldRun indicates whether the remote signer client should start
+	// fully or not.
+	shouldRun bool
+
+	// requestTimeout is the timeout used when sending responses to the
+	// watch-only node.
+	requestTimeout time.Duration
+
+	// retryTimeout is the backoff timeout used when retrying to set up a
+	// connection to the watch-only node, if the previous connection/attempt
+	// failed.
+	retryTimeout time.Duration
+
+	stopped int32 // To be used atomically.
+
+	quit chan struct{}
+
+	wg sync.WaitGroup
+
+	// wgMu ensures that we can't spawn a new Run goroutine after we've
+	// stopped the remote signer client.
+	wgMu sync.Mutex
+}
+
+// NewRemoteSignerClient creates a new instance of the remote signer client.
+// The passed subServers need to include a walletrpc.WalletKitServer and a
+// signrpc.SignerServer, or the RemoteSignerClient will be disabled.
+// Note that the client will only fully start if the configuration
+// enables an outbound remote signer.
+func NewRemoteSignerClient(subServers []lnrpc.SubServer,
+	streamFeeder SignCoordinatorStreamFeeder,
+	cfg *lncfg.RemoteSigner) *RemoteSignerClient {
+
+	var (
+		walletServer walletrpc.WalletKitServer
+		signerServer signrpc.SignerServer
+	)
+
+	for _, subServer := range subServers {
+		if server, ok := subServer.(walletrpc.WalletKitServer); ok {
+			walletServer = server
+		}
+
+		if server, ok := subServer.(signrpc.SignerServer); ok {
+			signerServer = server
+		}
+	}
+
+	if walletServer == nil {
```

Comment:
Ah, is this so we can always create a client on `lnd` startup but potentially not use it if it's a normal, standalone setup?
Perhaps it would be easier to understand (and IMO probably also a bit more compact in the code of this file), if we have a stateless function `ShouldCreateClient` that returns a bool, depending on these checks. Then only if that returns `true` do we even create a client.

Reply:
Perhaps we could return a `NoOpClient` if the function mentioned above returns `false`, so we can still start/stop it normally in the `lnd` startup flow.

Reply:
> Ah, is this so we can always create a client on lnd startup but potentially not use it if it's a normal, standalone setup?

Yeah that's correct, that's the reason for it.

> Perhaps we could return a NoOpClient if the function mentioned above returns false, so we can still start/stop it normally in the lnd startup flow.

Really good idea, things became a good bit cleaner by adding the `NoOpClient`. I mimicked the idea with of the `remote_signer_builder.go` though, and made a `remote_signer_client_builder.go` to either create the `NoOpClient` or that actual remote signer client. My reasoning behind that setup, is that it'll likely be quite a clean way to initialise the remote signer client, once we add optional validation for it.